Serialization of mask and background shorthands from computed style is incorrect

darinadler opened the pull request:
#### c9111c36f045cce09a0fe381727f4d3b9ddf08ca
<pre>
Serialization of mask and background shorthands from computed style is incorrect
<a href="https://bugs.webkit.org/show_bug.cgi?id=252460">https://bugs.webkit.org/show_bug.cgi?id=252460</a>
rdar://problem/105582183

Reviewed by NOBODY (OOPS!).

Use ShorthandSerializer for the background and mask shorthands even with computed and resolved style.
Eventually we will use it for all shorthands.

* LayoutTests/fast/backgrounds/background-shorthand-with-backgroundSize-style-expected.txt:
* LayoutTests/fast/backgrounds/background-shorthand-with-backgroundSize-style.html:
Updated to expect computed backgrounds to be serialized as specified. Also tweaked the test to be
a little bit easier to read and understand.

* LayoutTests/fast/dom/background-shorthand-csstext-expected.txt:
* LayoutTests/fast/dom/background-shorthand-csstext.html:
Updated to not expect the default value &quot;repeat&quot; to be included in shorthand serialization.

* LayoutTests/fast/masking/parsing-mask-expected.txt:
* LayoutTests/fast/masking/parsing-mask.html:
Updated to not expect default values to be included in resolved style shorthand serialization.

* LayoutTests/fast/masking/parsing-webkit-mask-expected.txt:
* LayoutTests/fast/masking/parsing-webkit-mask.html:
Updated to not expect the default value &quot;repeat&quot; to be included in shorthand serialization.

* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/style-src/inline-style-allowed-while-cloning-objects.sub-expected.txt:
Updated to not expect default values to be included in resolved style shorthand serialization.

* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-serialization/cssStyleValue-cssom.html:
Updated expected results to allow, but not require, background shorthand to serialize longhand
values in the order specified. I was a little surprised that other browsers are serializing in a
different order, and also that the typed OM computed style map does not serialize shorthands
following the &quot;shortest representation&quot; rule.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks-expected.txt:
Updated for progression now that mask and serialize it&apos;s initial resolved value.

* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-002-expected.txt:
Updated to expect two more PASS because of correct serialization of the resolved value of mask.

* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
Updated to expect two more PASS because of correct serialization of the resolved value of mask.

* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
Updated to expect a different failure now that correct serialization of the resolved value of
mask gets us past the earlier failure.

* LayoutTests/svg/css/getComputedStyle-basic-expected.txt: Updated types, because repeat values
are now CSSBackgroundRepeatValue, which is not a CSSPrimitiveValue. If we wanted to preserve the
old behavior in the legacy DOM we could, but I think we don&apos;t.

* Source/WebCore/animation/CSSPropertyAnimation.cpp: Removed unneeded include.

* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::getKeyframes): Updated for ComputedStyleExtractor changes.
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::commitStyles): Ditto.

* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
(WebCore::CSSComputedStyleDeclaration::CSSComputedStyleDeclaration): Take an options
argument instead of a boolean for allowVisitedStyle.
(WebCore::CSSComputedStyleDeclaration::create): Moved the comment about resolved values
rather than computed values here, since this is the function that passes in the
ProduceResolvedValues option. Updated to use options.
(WebCore::CSSComputedStyleDeclaration::createAllowingVisitedLinkColoring): Added.
(WebCore::CSSComputedStyleDeclaration::extractor const): Added.
(WebCore::CSSComputedStyleDeclaration::getPropertyCSSValue const): Use extractor.
(WebCore::CSSComputedStyleDeclaration::copyProperties const): Ditto.
(WebCore::CSSComputedStyleDeclaration::getPropertyValue const): Use
serializeShorthandValue rather than calling getPropertyCSSValue and cssText.
(WebCore::CSSComputedStyleDeclaration::getPropertyCSSValue): Use extractor.
(WebCore::CSSComputedStyleDeclaration::getPropertyValue): Ditto.

* Source/WebCore/css/CSSComputedStyleDeclaration.h: Removed unneeded includes.
Updated for above changes.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::createSingleAxisPositionValueForLayer): Use CSSValuePair so a single value
of this type is not misinterpreted as a list.
(WebCore::computedTransform): Use options instead of valueType.
(WebCore::adjustLengthForZoom): Removed AdjustPixelValuesForComputedStyle.
(WebCore::valueForShadow): Converted this to a non-member function. Removed
AdjustPixelValuesForComputedStyle.
(WebCore::ComputedStyleExtractor::valueForFilter): Removed AdjustPixelValuesForComputedStyle.
(WebCore::ComputedStyleExtractor::ComputedStyleExtractor): Use options.
(WebCore::fillRepeatToCSSValue): Use CSSBackgroundRepeatValue.
(WebCore::fillSizeToCSSValue): Use CSSValuePair so a single value of this type is not
misinterpreted as a list.
(WebCore::optionalLineHeight): Use options.
(WebCore::lineHeight): Ditto.
(WebCore::fontShorthandValue): Ditto.
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const): Ditto.
(WebCore::ComputedStyleExtractor::shouldUseLegacyShorthandSerialization): Added.
Returns true for all the untested shorthands. Eventually this will be deleted.

* Source/WebCore/css/ComputedStyleExtractor.h: Added ComputedStyleExtractorOption.
Removed WTF_MAKE_FAST_ALLOCATED since this is no longer allocated on the heap.
Made all member functions const.

* Source/WebCore/css/PropertySetCSSStyleDeclaration.cpp:
(WebCore::PropertySetCSSStyleDeclaration::getPropertyValue): Removed unneeded
isExposed check, already done by getPropertyValueInternal.

* Source/WebCore/css/SVGCSSComputedStyleDeclaration.cpp:
(WebCore::ComputedStyleExtractor::svgPropertyValue const): Made const.

* Source/WebCore/css/ShorthandSerializer.cpp: Added ShortestRepresentation
option.
(WebCore::ShorthandSerializer::ShorthandSerializer): Generalized to work
with either type of properties object.
(WebCore::ShorthandSerializer::commonSerializationChecks): Added a version
for ComputedStyleExtractor, which is much smipler than the StyleProperties one.
(WebCore::LayerValues::LayerValues): Added a flag for when to use the shortest
representation.
(WebCore::LayerValues::set): Implement shortest respresentation flag.
(WebCore::ShorthandSerializer::serializeLayered const): Ditto.
(WebCore::serializeShorthandValue): Added overlod for ComputedStyleExtractor.
(WebCore::serializeShorthandValueWithoutShortestRepresentationRule): Added.
* Source/WebCore/css/ShorthandSerializer.h: Updated for the above.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::isInitialValueForLonghand): Added a case for background-repeat and
mask-repeat, since they are stored as CSSBackgroundRepeatValue.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeBackgroundSize): Removed code that
tries to expand a single value to two. Instead, that is done in style building.

* Source/WebCore/css/typedom/ComputedStylePropertyMapReadOnly.cpp:
(WebCore::ComputedStylePropertyMapReadOnly::propertyValue const): Updated for
changes to ComputedStyleExtractor.
(WebCore::ComputedStylePropertyMapReadOnly::shorthandPropertySerialization const):
Use serializeShorthandValueWithoutShortestRepresentationRule.
(WebCore::ComputedStylePropertyMapReadOnly::customPropertyValue const): Updated
for chages to ComputedStyleExtractor.
(WebCore::ComputedStylePropertyMapReadOnly::entries const): Ditto.

* Source/WebCore/editing/ApplyStyleCommand.cpp: Added include.

* Source/WebCore/editing/EditingStyle.cpp: Added include.
(WebCore::copyPropertiesFromComputedStyle): Take const argument.

* Source/WebCore/editing/EditingStyle.h: Removed unneeded forward declaration.

* Source/WebCore/editing/ReplaceSelectionCommand.cpp: Added include.

* Source/WebCore/editing/cocoa/HTMLConverter.mm: Added include.
(HTMLConverterCaches::computedStylePropertyForElement): Use a new
ComputedStyleExtractor each time since it&apos;s a lightweight object.

* Source/WebCore/html/CustomPaintImage.cpp:
(WebCore::extractComputedProperty): Updated for chagnes to ComputedStyleExtractor.
* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
(WebCore::buildObjectForKeyframes): Ditto.

* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::getComputedStyleForNode): Use
CSSComputedStyleDeclaration::createAllowingVisitedLinkColoring.

* Source/WebCore/page/DOMWindow.cpp:
(WebCore::DOMWindow::getComputedStyle const): Updated for changes to
ComputedStyleExtractor.
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::changedCustomPaintWatchedProperty): Ditto.
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::SVGElement::colorInterpolation const): Ditto.

* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::computedStyleIncludingVisitedInfo const):
Use CSSComputedStyleDeclaration::createAllowingVisitedLinkColoring.

* Source/WebCore/testing/Internals.h: Added forward declaration of Node.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c9111c36f045cce09a0fe381727f4d3b9ddf08ca

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108491 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17587 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41340 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117596 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/117801 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112377 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19038 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8865 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100718 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114258 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/14269 "Found 16 new test failures: fast/backgrounds/size/parsing-background-size-values.html, fast/backgrounds/size/parsing-inherit.html, fast/css/background-layers-initial-size.html, fast/css/cascade/background-size-and-webkit-background-size-cascade-order.html, fast/css/getComputedStyle/getComputedStyle-background-shorthand.html, fast/css/getComputedStyle/getComputedStyle-background-size.html, fast/css/getComputedStyle/getComputedStyle-zoom-and-background-size.html, fast/css/value-list-out-of-bounds-crash.html, imported/w3c/web-platform-tests/css/css-animations/KeyframeEffect-getKeyframes.tentative.html, imported/w3c/web-platform-tests/css/css-backgrounds/background-332.html ... (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97490 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42229 "Found 17 new test failures: editing/pasteboard/5780697-2.html, fast/backgrounds/size/parsing-background-size-values.html, fast/backgrounds/size/parsing-inherit.html, fast/css/background-layers-initial-size.html, fast/css/cascade/background-size-and-webkit-background-size-cascade-order.html, fast/css/getComputedStyle/getComputedStyle-background-shorthand.html, fast/css/getComputedStyle/getComputedStyle-background-size.html, fast/css/getComputedStyle/getComputedStyle-zoom-and-background-size.html, fast/css/value-list-out-of-bounds-crash.html, imported/w3c/web-platform-tests/css/css-animations/KeyframeEffect-getKeyframes.tentative.html ... (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/96247 "Passed tests") | [⏳ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/macOS-BigSur-Release-WK1-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/83938 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10398 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30479 "Found 17 new test failures: editing/pasteboard/5780697-2.html, fast/backgrounds/size/parsing-background-size-values.html, fast/backgrounds/size/parsing-inherit.html, fast/css/background-layers-initial-size.html, fast/css/cascade/background-size-and-webkit-background-size-cascade-order.html, fast/css/getComputedStyle/getComputedStyle-background-shorthand.html, fast/css/getComputedStyle/getComputedStyle-background-size.html, fast/css/getComputedStyle/getComputedStyle-zoom-and-background-size.html, fast/css/value-list-out-of-bounds-crash.html, imported/w3c/web-platform-tests/css/css-animations/KeyframeEffect-getKeyframes.tentative.html ... (failure)") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11154 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7392 "Found 17 new test failures: editing/pasteboard/5780697-2.html, fast/backgrounds/size/parsing-background-size-values.html, fast/backgrounds/size/parsing-inherit.html, fast/css/background-layers-initial-size.html, fast/css/cascade/background-size-and-webkit-background-size-cascade-order.html, fast/css/getComputedStyle/getComputedStyle-background-shorthand.html, fast/css/getComputedStyle/getComputedStyle-background-size.html, fast/css/getComputedStyle/getComputedStyle-zoom-and-background-size.html, fast/css/value-list-out-of-bounds-crash.html, imported/w3c/web-platform-tests/css/css-animations/KeyframeEffect-getKeyframes.tentative.html ... (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16544 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50075 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12742 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c816e47becbf24d35576a18041112c3b4041dc64. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10323

darinadler committed:
Serialization of mask and background shorthands from computed style is incorrect
https://bugs.webkit.org/show_bug.cgi?id=252460
rdar://problem/105582183

Reviewed by NOBODY (OOPS!).

Use ShorthandSerializer for the background and mask shorthands even with computed and resolved style.
Eventually we will use it for all shorthands.

* LayoutTests/fast/backgrounds/background-shorthand-with-backgroundSize-style-expected.txt:
* LayoutTests/fast/backgrounds/background-shorthand-with-backgroundSize-style.html:
Updated to expect computed backgrounds to be serialized as specified. Also tweaked the test to be
a little bit easier to read and understand.

* LayoutTests/fast/dom/background-shorthand-csstext-expected.txt:
* LayoutTests/fast/dom/background-shorthand-csstext.html:
Updated to not expect the default value "repeat" to be included in shorthand serialization.

* LayoutTests/fast/masking/parsing-mask-expected.txt:
* LayoutTests/fast/masking/parsing-mask.html:
Updated to not expect default values to be included in resolved style shorthand serialization.

* LayoutTests/fast/masking/parsing-webkit-mask-expected.txt:
* LayoutTests/fast/masking/parsing-webkit-mask.html:
Updated to not expect the default value "repeat" to be included in shorthand serialization.

* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/style-src/inline-style-allowed-while-cloning-objects.sub-expected.txt:
Updated to not expect default values to be included in resolved style shorthand serialization.

* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-serialization/cssStyleValue-cssom.html:
Updated expected results to allow, but not require, background shorthand to serialize longhand
values in the order specified. I was a little surprised that other browsers are serializing in a
different order, and also that the typed OM computed style map does not serialize shorthands
following the "shortest representation" rule.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks-expected.txt:
Updated for progression now that mask and serialize it's initial resolved value.

* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-002-expected.txt:
Updated to expect two more PASS because of correct serialization of the resolved value of mask.

* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
Updated to expect two more PASS because of correct serialization of the resolved value of mask.

* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
Updated to expect a different failure now that correct serialization of the resolved value of
mask gets us past the earlier failure.

* LayoutTests/svg/css/getComputedStyle-basic-expected.txt: Updated types, because repeat values
are now CSSBackgroundRepeatValue, which is not a CSSPrimitiveValue. If we wanted to preserve the
old behavior in the legacy DOM we could, but I think we don't.

* Source/WebCore/animation/CSSPropertyAnimation.cpp: Removed unneeded include.

* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::getKeyframes): Updated for ComputedStyleExtractor changes.
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::commitStyles): Ditto.

* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
(WebCore::CSSComputedStyleDeclaration::CSSComputedStyleDeclaration): Take an options
argument instead of a boolean for allowVisitedStyle.
(WebCore::CSSComputedStyleDeclaration::create): Moved the comment about resolved values
rather than computed values here, since this is the function that passes in the
ProduceResolvedValues option. Updated to use options.
(WebCore::CSSComputedStyleDeclaration::createAllowingVisitedLinkColoring): Added.
(WebCore::CSSComputedStyleDeclaration::extractor const): Added.
(WebCore::CSSComputedStyleDeclaration::getPropertyCSSValue const): Use extractor.
(WebCore::CSSComputedStyleDeclaration::copyProperties const): Ditto.
(WebCore::CSSComputedStyleDeclaration::getPropertyValue const): Use
serializeShorthandValue rather than calling getPropertyCSSValue and cssText.
(WebCore::CSSComputedStyleDeclaration::getPropertyCSSValue): Use extractor.
(WebCore::CSSComputedStyleDeclaration::getPropertyValue): Ditto.

* Source/WebCore/css/CSSComputedStyleDeclaration.h: Removed unneeded includes.
Updated for above changes.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::createSingleAxisPositionValueForLayer): Use CSSValuePair so a single value
of this type is not misinterpreted as a list.
(WebCore::computedTransform): Use options instead of valueType.
(WebCore::adjustLengthForZoom): Removed AdjustPixelValuesForComputedStyle.
(WebCore::valueForShadow): Converted this to a non-member function. Removed
AdjustPixelValuesForComputedStyle.
(WebCore::ComputedStyleExtractor::valueForFilter): Removed AdjustPixelValuesForComputedStyle.
(WebCore::ComputedStyleExtractor::ComputedStyleExtractor): Use options.
(WebCore::fillRepeatToCSSValue): Use CSSBackgroundRepeatValue.
(WebCore::fillSizeToCSSValue): Use CSSValuePair so a single value of this type is not
misinterpreted as a list.
(WebCore::optionalLineHeight): Use options.
(WebCore::lineHeight): Ditto.
(WebCore::fontShorthandValue): Ditto.
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const): Ditto.
(WebCore::ComputedStyleExtractor::shouldUseLegacyShorthandSerialization): Added.
Returns true for all the untested shorthands. Eventually this will be deleted.

* Source/WebCore/css/ComputedStyleExtractor.h: Added ComputedStyleExtractorOption.
Removed WTF_MAKE_FAST_ALLOCATED since this is no longer allocated on the heap.
Made all member functions const.

* Source/WebCore/css/PropertySetCSSStyleDeclaration.cpp:
(WebCore::PropertySetCSSStyleDeclaration::getPropertyValue): Removed unneeded
isExposed check, already done by getPropertyValueInternal.

* Source/WebCore/css/SVGCSSComputedStyleDeclaration.cpp:
(WebCore::ComputedStyleExtractor::svgPropertyValue const): Made const.

* Source/WebCore/css/ShorthandSerializer.cpp: Added ShortestRepresentation
option.
(WebCore::ShorthandSerializer::ShorthandSerializer): Generalized to work
with either type of properties object.
(WebCore::ShorthandSerializer::commonSerializationChecks): Added a version
for ComputedStyleExtractor, which is much smipler than the StyleProperties one.
(WebCore::LayerValues::LayerValues): Added a flag for when to use the shortest
representation.
(WebCore::LayerValues::set): Implement shortest respresentation flag.
(WebCore::ShorthandSerializer::serializeLayered const): Ditto.
(WebCore::serializeShorthandValue): Added overlod for ComputedStyleExtractor.
(WebCore::serializeShorthandValueWithoutShortestRepresentationRule): Added.
* Source/WebCore/css/ShorthandSerializer.h: Updated for the above.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::isInitialValueForLonghand): Added a case for background-repeat and
mask-repeat, since they are stored as CSSBackgroundRepeatValue.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeBackgroundSize): Removed code that
tries to expand a single value to two. Instead, that is done in style building.

* Source/WebCore/css/typedom/ComputedStylePropertyMapReadOnly.cpp:
(WebCore::ComputedStylePropertyMapReadOnly::propertyValue const): Updated for
changes to ComputedStyleExtractor.
(WebCore::ComputedStylePropertyMapReadOnly::shorthandPropertySerialization const):
Use serializeShorthandValueWithoutShortestRepresentationRule.
(WebCore::ComputedStylePropertyMapReadOnly::customPropertyValue const): Updated
for chages to ComputedStyleExtractor.
(WebCore::ComputedStylePropertyMapReadOnly::entries const): Ditto.

* Source/WebCore/editing/ApplyStyleCommand.cpp: Added include.

* Source/WebCore/editing/EditingStyle.cpp: Added include.
(WebCore::copyPropertiesFromComputedStyle): Take const argument.

* Source/WebCore/editing/EditingStyle.h: Removed unneeded forward declaration.

* Source/WebCore/editing/ReplaceSelectionCommand.cpp: Added include.

* Source/WebCore/editing/cocoa/HTMLConverter.mm: Added include.
(HTMLConverterCaches::computedStylePropertyForElement): Use a new
ComputedStyleExtractor each time since it's a lightweight object.

* Source/WebCore/html/CustomPaintImage.cpp:
(WebCore::extractComputedProperty): Updated for chagnes to ComputedStyleExtractor.
* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
(WebCore::buildObjectForKeyframes): Ditto.

* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::getComputedStyleForNode): Use
CSSComputedStyleDeclaration::createAllowingVisitedLinkColoring.

* Source/WebCore/page/DOMWindow.cpp:
(WebCore::DOMWindow::getComputedStyle const): Updated for changes to
ComputedStyleExtractor.
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::changedCustomPaintWatchedProperty): Ditto.
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::SVGElement::colorInterpolation const): Ditto.

* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::computedStyleIncludingVisitedInfo const):
Use CSSComputedStyleDeclaration::createAllowingVisitedLinkColoring.

* Source/WebCore/testing/Internals.h: Added forward declaration of Node.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c9111c36f045cce09a0fe381727f4d3b9ddf08ca. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10323

webkit-early-warning-system commented:
EWS run on previous version of this PR (hash https://github.com/WebKit/WebKit/commit/9fb5e3e067d460a8a765d590795ac354a8b94429)<details>

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108490 "Passed style check") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17586 "Build is in progress. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41339 "Build is in progress. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117595 "Build is in progress. Recent messages:Failed to print configuration; Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Updated wpe dependencies; Running compile-webkit") | [⏳ 🛠 wincairo ](https://ews-build.webkit.org/#/builders/WinCairo-EWS "Waiting in queue, processing has not started yet") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112376 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19037 "Build is in progress. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8864 "Build is in progress. Recent messages:Pull request contains relevant changes; Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings); Running upload-built-product") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100717 "Build is in progress. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Updated gtk dependencies; Running compile-webkit") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114257 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/76/builds/19037 "Build is in progress. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/43/builds/41339 "Build is in progress. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/36/builds/100717 "Build is in progress. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Updated gtk dependencies; Running compile-webkit") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/76/builds/19037 "Build is in progress. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/43/builds/41339 "Build is in progress. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/36/builds/100717 "Build is in progress. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Updated gtk dependencies; Running compile-webkit") | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-16-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/43/builds/41339 "Build is in progress. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/85/builds/8864 "Build is in progress. Recent messages:Pull request contains relevant changes; Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings); Running upload-built-product") | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 services ](https://ews-build.webkit.org/#/builders/Services-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-9-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/43/builds/41339 "Build is in progress. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [⏳ 🛠 jsc-mips ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12741 "Built successfully") | | [⏳ 🧪 jsc-mips-tests ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | | | | </details>

graouts commented on LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
```diff
@@ -58,9 +58,9 @@ PASS marker-start uses discrete animation when animating between "url("http://lo
-FAIL mask uses discrete animation when animating between "url("http://localhost/test-1")" and "url("http://localhost/test-2")" with linear easing assert_equals: The value should be url("http://localhost/test-1") at 0ms expected "url(\"http://localhost/test-1\")" but got "url(\"http://localhost/test-1\") 0% 0% / auto repeat border-box border-box add match-source"
-FAIL mask uses discrete animation when animating between "url("http://localhost/test-1")" and "url("http://localhost/test-2")" with effect easing assert_equals: The value should be url("http://localhost/test-1") at 0ms expected "url(\"http://localhost/test-1\")" but got "url(\"http://localhost/test-1\") 0% 0% / auto repeat border-box border-box add match-source"
-FAIL mask uses discrete animation when animating between "url("http://localhost/test-1")" and "url("http://localhost/test-2")" with keyframe easing assert_equals: The value should be url("http://localhost/test-1") at 0ms expected "url(\"http://localhost/test-1\")" but got "url(\"http://localhost/test-1\") 0% 0% / auto repeat border-box border-box add match-source"
+FAIL mask uses discrete animation when animating between "url("http://localhost/test-1")" and "url("http://localhost/test-2")" with linear easing assert_equals: The value should be url("http://localhost/test-1") at 499ms expected "url(\"http://localhost/test-1\")" but got "url(\"http://localhost/test-2\")"
+FAIL mask uses discrete animation when animating between "url("http://localhost/test-1")" and "url("http://localhost/test-2")" with effect easing assert_equals: The value should be url("http://localhost/test-1") at 940ms expected "url(\"http://localhost/test-1\")" but got "url(\"http://localhost/test-2\")"
+FAIL mask uses discrete animation when animating between "url("http://localhost/test-1")" and "url("http://localhost/test-2")" with keyframe easing assert_equals: The value should be url("http://localhost/test-1") at 940ms expected "url(\"http://localhost/test-1\")" but got "url(\"http://localhost/test-2\")"
```
Not good. The correct value for `mask-image` was picked up prior to this change, but not anymore. This looks like interpolation does not happen at all. Yet, `background-image` which shares the same interpolation code (`FillLayerStyleImagePropertyWrapper`) does not seem to suffer that fate. Not obvious to me why just looking at the code.

graouts reviewed (changes requested):
I think this bit in your ChangeLog is incorrect:

> * LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
Updated to expect a different failure now that correct serialization of the resolved value of
mask gets us past the earlier failure.

The correct value for mask-image was picked up prior to this change and only serialization was the issue, but not anymore. This looks like interpolation does not happen at all. Yet, background-image which shares the same interpolation code (`FillLayerStyleImagePropertyWrapper`) does not seem to suffer that fate. Not obvious to me why just looking at the code.

graouts commented:
> I think this bit in your ChangeLog is incorrect:
> 
> > * LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
> >   Updated to expect a different failure now that correct serialization of the resolved value of
> >   mask gets us past the earlier failure.
> 
> The correct value for mask-image was picked up prior to this change and only serialization was the issue, but not anymore. This looks like interpolation does not happen at all. Yet, background-image which shares the same interpolation code (`FillLayerStyleImagePropertyWrapper`) does not seem to suffer that fate. Not obvious to me why just looking at the code.

Actually, I take that back. We would fail right at the first of many tests with no indication that we correctly blended `mask-image`. I take back my request for changes.

darinadler commented:
Most likely we will solve this problem a different way; the work that @weinig is doing to reorganize style code may make the fix more straightforward.
